File: problems/14_2d_dp/0062_unique_paths/solution.swift
// 0062. Unique Paths
import Foundation
class Solution {
    func uniquePaths(_ m: Int, _ n: Int) -> Int {
        var dp = Array(repeating: 1, count: n)
        if m > 1 {
            for _ in 1..<m { for j in 1..<n { dp[j] += dp[j - 1] } }
        }
        return dp[n - 1]
    }
}
let sol = Solution()
if sol.uniquePaths(3, 7) != 28 { fatalError("a") }
if sol.uniquePaths(3, 2) != 3 { fatalError("b") }
if sol.uniquePaths(7, 3) != 28 { fatalError("c") }
if sol.uniquePaths(1, 1) != 1 { fatalError("d") }
if sol.uniquePaths(1, 10) != 1 { fatalError("e") }
if sol.uniquePaths(10, 10) != 48620 { fatalError("f") }
print("All tests passed!")

File: problems/14_2d_dp/0062_unique_paths/solution.dart
// 0062. Unique Paths
int uniquePaths(int m, int n) {
  final dp = List<int>.filled(n, 1);
  for (var i = 1; i < m; i++) {
    for (var j = 1; j < n; j++) dp[j] += dp[j - 1];
  }
  return dp[n - 1];
}
void main() {
  if (uniquePaths(3, 7) != 28) throw 'a';
  if (uniquePaths(3, 2) != 3) throw 'b';
  if (uniquePaths(7, 3) != 28) throw 'c';
  if (uniquePaths(1, 1) != 1) throw 'd';
  if (uniquePaths(1, 10) != 1) throw 'e';
  if (uniquePaths(10, 10) != 48620) throw 'f';
  print('All tests passed!');
}
